### PR TITLE
Implement block header

### LINKED ISSUE #32: Block Header Implementation

### PR DESCRIPTION
**Pull request by rtsisyk**
- Add stubs for PublicKey, Signature, Payload and other
- Define Input, Output, BlockHeader and Block structures
- Implement block signature
- Implement genesis block in simplified way

Closes #32

### CODE REVIEW

**Comment by dbmcclain** [comment]
I implemented needed items in `crypto::pbc`
1. New dummy constructor for `crypto::pbc::secure::RVal` used in encrypted payloads, printable, hash-able
2. Implement `crypto::pbc::secure::Signature` type - printable, hash-able
3. make internal routines Hash::from_vector() and Hash::from_str() both utilize the new Hasher object and Hashable.
4. Make crypto::Hash items re-hash-able
5. Implement error handling where needed
6. Base our hex strings around the hex crate functions

**Comment by dbmcclain** [comment]
Not sure what you are thinking here about EncryptedPayload. That is not a crypto function, but something specific to a particular application.

The pbc::secure module defines a complete structure called EncryptedPacket that includes a public key, an ID tag, the RVal, and the crypto-text. This is a generic SAKKI encrypted output that can be used by applications as they see fit.

Our blockchain is just one particular application for SAKKI encryption, and it only needs to show the RVal and crypto-text portions. The public key shall remain hidden, known already by the intended recipient, and also same for the ID tag.

I argue that this kind of specificity for blockchain does not belong in a general purpose crypto library, and so the definition of EncryptedPayload belongs with the client code that cherry picks what it needs from a SAKKI EncryptedPacket structure.

**Comment by rtsisyk** [comment]
> Not sure what you are thinking here about EncryptedPayload. That is not a crypto function, but something specific to a particular application.

OK, I agree with you. I moved EncryptedPayload back to blockchain library. Definition is from your comment above.

> As for the missing libPBC in Linux... I am at a loss to help you there. Why did it work previously and not now?

This problem was caused by a bug in our CI scripts. This has been already fixed.

Rebased and force-pushed. Please review.

File: blockchain/src/output.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            recipient: recipient,
+            proof: proof,
+            payload: payload,
+        }
+    }
+}
+
+impl Hashable for Output {
+    fn hash(&self, state: &mut Hasher) {
+        // Don't include self.hash because it is redundant in this case
```
Comment (dbmcclain) [inline blockchain/src/output.rs]: ... self.hash not just redundant - ID should equal the result of this call to hash. You might want to check that condition to be sure you aren't dealing with garbage data from some forgery.

File: crypto/src/lib.rs [hunk 2]
```diff
@@ -25,17 +25,42 @@ extern crate hex;
...
 
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+/// Stub for Bullet Proof
+// TODO: define
+#[derive(Clone)]
+pub struct BulletProof([u8; 4096]);
```
Comment (dbmcclain) [inline crypto/src/lib.rs]: yes, as good a guess as any. Should flag this with TODO to come back and fix once we know what a Bulletproof actually looks like. (Actually, I'm the one that will likely do this...)

File: blockchain/src/block.rs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    /// Merklish root of all range proofs for output.
+    pub outputs_range_hash: Hash,
+
+    /// Ordered list of witness public keys for current epoch.
+    /// The leader node is also considered a witness during the current epoch.
+    pub witnesses: Vec<PublicKey>,
+
+    /// CoSi multisignature on HCURR
+    // TODO: which kind
+    pub sig: G1,
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply blockchain/src/block.rs]: OK

File: blockchain/src/block.rs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+impl Block {
+    pub fn sign(
+        version: u64,
+        epoch: u64,
+        previous: Hash,
+        leader: PublicKey,
+        adjustment: Zr,
+        witnesses: &[PublicKey],
+        sig: G1,
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply blockchain/src/block.rs]: OK

File: blockchain/src/block.rs [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        timestamp.hash(&mut hasher);
+        inputs_range_hash.hash(&mut hasher);
+        outputs_range_hash.hash(&mut hasher);
+        // hash the number of witnesses first
+        let witnesses_count: u64 = witnesses.len() as u64;
+        witnesses_count.hash(&mut hasher);
+        for witness in witnesses.iter() {
+            witness.hash(&mut hasher);
+        }
+        sig.hash(&mut hasher);
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply blockchain/src/block.rs]: OK, I understood the idea. I'll rework this part.

File: blockchain/src/block.rs [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        let keys = make_deterministic_keys(&seed);
+
+        let leader = keys.1;
+
+        let version: u64 = 1;
+        let epoch: u64 = 1;
+        let previous = Hash::from_str(GENESIS_BLOCK_HASH_STRING);
+        let adjustment: Zr = Zr::new();
+        let witnesses = [leader.clone()];
+        let sig: G1 = G1::new();
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply blockchain/src/block.rs]: I'll rework it today.

File: blockchain/src/block.rs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+        // But have one hard-coded output
+        let proof = BulletProof::garbage();
+        let payload = EncryptedPayload::garbage();
+        let output = Output::new(leader.clone(), proof, payload);
+        let outputs = [output];
+
+        Block::sign(
+            version, epoch, previous, leader, adjustment, &witnesses, sig, &inputs, &outputs,
+        )
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply blockchain/src/block.rs]: OK

File: blockchain/src/input.rs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+use stegos_crypto::hash::{Hash, Hashable, Hasher};
+use stegos_crypto::pbc::secure::Signature;
+
+/// Transaction Input.
+#[derive(Debug, Clone)]
+pub struct Input {
+    /// Identifier of an unspent transaction output.
+    pub source_id: Hash,
+    /// Signature used to sign transactions.
+    pub signature: Signature,
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply blockchain/src/input.rs]: I guessed that Input should be signed by sender...

File: blockchain/src/output.rs (hunk 1 above)
Comment (rtsisyk) [reply blockchain/src/output.rs]: This function is used when Output itself is hashed, i.e. for create the block signature.
There are no difference between hash(hash(a, b, c), a, b, c) and hash(a, b, c) in terms of signing.

**Comment by rtsisyk** [comment]
I'm reworking it according to David's comments.

File: blockchain/src/output.rs (hunk 1 above)
Comment (dbmcclain) [reply blockchain/src/output.rs]: yes... there really is a difference... but maybe you don't need to care?

**Pull request merged**